src/jit/codegenxarch.cpp

```diff
@@ -6841,37 +6908,79 @@ void        CodeGen::genCompareLong(GenTreePtr  treeNode)
+    
+    // TODO-X86-CQ: When targetReg == REG_NA, if the next block is a BBJ_ALWAYS, we can set
+    //              labelFinal = compiler->compCurBB->bbNext->bbJumpDest.
+    BasicBlock* labelHi = genCreateTempLabel();
+    BasicBlock* labelFinal = genCreateTempLabel();
 
-    // Generate the second jump for LE, LT, GT, and GE. We only do the lower compare if
-    // the hi parts are equal
-    if (jumpKind[1] != EJ_NONE)
+    if (compareKind == CK_SIGNED && (tree->gtOper != GT_NE && tree->gtOper != GT_EQ))
     {
-        assert(branchToTrueLabel[1]);
-        inst_JMP(jumpKind[1], label);
-    }
+        // If we are doing a signed comparison, we need to do a signed set if the high compare is true,
+        // but an unsigned set if we fall through to the low compare. If we have a GT_NE or GT_EQ, we do not
+        // need to worry about the sign of the comparison, so we can use the simplified case.
 
-    // Now create compare for low parts
-    ins = INS_cmp;
-    cmpType = TYP_INT;
-    cmpAttr = emitTypeSize(cmpType);
+        // We only have to check for equality for the hi comparison. If they are not equal, then the set will
+        // do the right thing. If they are equal, we have to check the lo halves.
+        inst_JMP(EJ_jne, labelHi);
 
-    // Emit the comparison
-    getEmitter()->emitInsBinary(ins, cmpAttr, loOp1, loOp2);
+        // Now create compare for low parts
+        ins = INS_cmp;
+        cmpType = TYP_INT;
+        cmpAttr = emitTypeSize(cmpType);
 
-    // Define the label for hi jump target here. If we have jumped here, we want to set
-    // the target register based on the jump kind of the lower half (the actual compare
-    // type). If we have fallen through, then we are doing a normal int compare for the
-    // lower parts
+        // Emit the comparison. Perform the set for the lo. Jump to labelFinal
+        getEmitter()->emitInsBinary(ins, cmpAttr, loOp1, loOp2);
 
-    genDefineTempLabel(label);
-    if (targetReg != REG_NA)
+        emitJumpKind jumpKindLo[2];
+        bool branchToTrueLabelLo[2];
+
+        // The low set must be unsigned
+        genJumpKindsForTreeLongLo(tree, jumpKindLo, branchToTrueLabelLo);
+        inst_SET(jumpKindLo[0], targetReg);
+        
+        // Set the higher bytes to 0
+        inst_RV_RV(ins_Move_Extend(TYP_UBYTE, true), targetReg, targetReg, TYP_UBYTE, emitTypeSize(TYP_UBYTE));
+        genProduceReg(tree);
+        inst_JMP(EJ_jmp, labelFinal);
+
+        // Define the label for hi jump target here. If we have jumped here, we want to set
+        // the target register based on the jump kind of the actual compare type.
+
+        genDefineTempLabel(labelHi);
+        inst_SET(genJumpKindForOper(tree->gtOper, compareKind), targetReg);
+
+        // Set the higher bytes to 0
+        inst_RV_RV(ins_Move_Extend(TYP_UBYTE, true), targetReg, targetReg, TYP_UBYTE, emitTypeSize(TYP_UBYTE));
+        genProduceReg(tree);
+
+        genDefineTempLabel(labelFinal);
+    }
```

Comment:
Couldn't you put the labelFinal above the zero extend? Now, you end up with two identical zero extends on different paths.